### PR TITLE
Create a new method : startPlayerFromBuffer, to play from a buffer

### PR DESCRIPTION
**Pull request by Larpoux**
- Create a new method : 'startPlayerFromBuffer' to play from a memory buffer instead of a file. 
(Tested on iOS and Android).
- Bug Fix: Fixes returned filePath on iOS #168

### CODE REVIEW

**Comment by Larpoux** [comment]
Hi  Hyochan.
I merged your fix into my fork.
Yes, the example runs correctly on iOS and Android.

**Comment by Larpoux** [comment]
Add support for OPUS encoder on iOS.
-----------------------------------
Actually on iOS, you can choose from two encoders :
- AAC (this is the default)
- CAF/OPUS

To encode with OPUS you do the following :
```dart
await flutterSound.startRecorder(null, codec: t_CODEC.CODEC_CAF_OPUS,);
```
Recently, Apple added a support for encoding with the standard OPUS codec. Unfortunately, Apple encapsulates its data in its own proprietary envelope : CAF. This is really stupid, this is Apple

On Android the OPUS codec is not yet supported by flutter_sound.
 master

**Comment by Larpoux** [comment]
Hi Hyochan,
I appreciate that you spend some time to review my pull request.

1. Your crash in example seems to be tied with using an old plugin instead of the one from my fork. Of course there is an implementation for method "startPlayerFromBuffer" in my fork. You should verify your dependencies/dev_dependencies in pubspec.yaml

2. I have no ideas what is i18n. It seems that Flutter or Android Studio add those files automatically. If necessary I will remove those files from the git repository.

3. I found a little bug in the example (not flutter_sound) when the user try to play from buffer but has not recorded anything previously. I pushed the fix in my fork

4. I forgot to mentioned that since AndroidSDK-29 the directory `getExternalStorageDirectory()` is not writable anymore. I changed the two calls to this function by `Environment.getDataDirectory()`. I did not modified the README.md accordingly but I will do it now.

5. Actually, the only encoder compatible with iOS and Android is AAC. My plan is to support correctly the regular/official VORBIS and OPUS encoders if not too difficult. (I need OPUS in my own App)

6. I noticed Request #154 : "Get audio samples instead of writing to a temp file #154". Maybe you can affect me on this request and I will report what is possible to do

**Comment by hyochan** [comment]
@Larpoux Thanks for your time for investigating our plugin. I've checked out newer commit and seems to be working correctly. Before any further review, I want to confirm few things.

From question 2, I've meant about localization file that automatically generates from `android studio`.
image
You seem to use the above plugin so that it generates the below file (It is just on of them).
image

> If you don't want to bother yourself on removing it, just keep it then I'll remove in the future when your `PR` has been merged.

Are you happy with the current `PR` to be merged to `master` branch? Then I'll take a look again then try to merge your `PR` if it looks ok.

**Comment by Larpoux** [comment]
Hi @hyochan ,

- I removed i18n from my plugins and did a push. Thank you for your help
- Yes, I am happy with my contribution, and will be glad if you validate it and integrate in the main master repository
- What about  Request #154 : "Get audio samples instead of writing to a temp file #154" ? Do you have comments on it ?

**Comment by hyochan** [comment]
@Larpoux I think #154 can be handled in another `PR`.

File: example/lib/main.dart [hunk 1]
```diff
@@ -101,11 +118,45 @@ class _MyAppState extends State<MyApp> {
...
+
   void startPlayer() async{
     try {
-      String path = await flutterSound.startPlayer(this._path);
-      await flutterSound.setVolume(1.0);
-      print('startPlayer: $path');
+      String path = null;
+      if (_playFromBuffer == 0) // Do we want to play from buffer or from file ?
+      {
+        path = await flutterSound.startPlayer (this._path); // From file
```
Comment (hyochan) [inline example/lib/main.dart]: Please remove space after `startPlayer`.

File: example/lib/main.dart [hunk 2]
```diff
@@ -101,11 +118,45 @@ class _MyAppState extends State<MyApp> {
...
+    }
+  }
+
   void startPlayer() async{
     try {
-      String path = await flutterSound.startPlayer(this._path);
-      await flutterSound.setVolume(1.0);
-      print('startPlayer: $path');
+      String path = null;
+      if (_playFromBuffer == 0) // Do we want to play from buffer or from file ?
```
Comment (hyochan) [inline example/lib/main.dart]: The indentation can go upward as described in [effective dart](https://dart.dev/guides/language/effective-dart/design#do-annotate-when-dart-infers-the-wrong-type).

Below is common format.
image

File: example/lib/main.dart (hunk 1 above)
Comment (Larpoux) [reply example/lib/main.dart]: done. Pushed.

File: example/lib/main.dart [hunk 3]
```diff
@@ -101,11 +118,45 @@ class _MyAppState extends State<MyApp> {
...
+      {
+        path = await flutterSound.startPlayer (this._path); // From file
+
+      } else
+        {
+          Uint8List buffer = await makeBuffer(this._path);
+          if (buffer != null)
+            path = await flutterSound.startPlayerFromBuffer (buffer); // From buffer
```
Comment (hyochan) [inline example/lib/main.dart]: same here.

File: example/lib/main.dart [hunk 4]
```diff
@@ -101,11 +118,45 @@ class _MyAppState extends State<MyApp> {
...
+        }
+      if (path == null)
+        {
+          print ('Error starting player');
+          return;
+        } else
```
Comment (hyochan) [inline example/lib/main.dart]: You've already returned it. Therefore, `else` is redundant.

```
if (path == null) {
  print ('Error starting player');
  return;
}
```
will do it

File: res/values/strings_en.arb [hunk 5]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (hyochan) [inline res/values/strings_en.arb]: Please remove current file either!

**Comment by hyochan** [comment]
@Larpoux I am happy with your `PR` except few comments I've provided!

**Comment by Larpoux** [comment]
Heah ! :-D.   
Thank you, I am currently fixing your remarks.

**Comment by Larpoux** [comment]
OK Hyo,
I think it's now better.
I must confess you:
I am a Pascal :-) programmer. I have much difficulties to read code when 'BEGIN' 'END' are not indented same ( '{' and '}' in C).  But things are changing and left '{' is not anymore on a specific line. Hard for me, but I agree that code must be homogeneous.

**Comment by hyochan** [comment]
> [effective dart documentation](https://dart.dev/guides/language/effective-dart/documentation).
@Larpoux How about this?

**Comment by Larpoux** [comment]
Thank you Hyo. 
- I will add soon some comments (///) for this new function
- I will be proud to be consider as a flutter_sound contributor
- I will work now on a new 'dev' branch. I am sorry, but I was not very fluent with github.

**Comment by hyochan** [comment]
@Larpoux No worry. You did great~! You may refer to [naming branches](https://github.com/agis/git-style-guide) as well.

**Pull request merged**